staging/src/k8s.io/client-go/tools/watch/until.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watch
+
+import (
+	"errors"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ConditionFunc returns true if the condition has been reached, false if it has not been reached yet,
+// or an error if the condition cannot be checked and should terminate. In general, it is better to define
+// level driven conditions over edge driven conditions (pod has ready=true, vs pod modified and ready changed
+// from false to true).
+type ConditionFunc func(event watch.Event) (bool, error)
+
+// ErrWatchClosed is returned when the watch channel is closed before timeout in UntilWithoutRetry.
+var ErrWatchClosed = errors.New("watch closed before UntilWithoutRetry timeout")
+
+// UntilWithoutRetry reads items from the watch until each provided condition succeeds, and then returns the last watch
+// encountered. The first condition that returns an error terminates the watch (and the event is also returned).
+// If no event has been received, the returned event will be nil.
+// Conditions are satisfied sequentially so as to provide a useful primitive for higher level composition.
+// A zero timeout means to wait forever.
+//
+// Warning: Unless you have a very specific use case (probably a special Watcher) don't use this function!!!
```

Comment:
Do you see any usecase for it?

Why do we even make it public then?

Reply:
I agree with @wojtek-t. If we expose a function, someone will eventually use it.

Reply:
I we keep it we don't have to switch all the places at once. There are also many cases not covered by this PR, temporarily switched to pooling, that might raise the need. (Like creating the watcher in advance at some specific point to capture RV, although you can do it with List - unless you have RBAC only for Watch.) I don't feel strong about it though.

Reply:
If that's about making this PR smaller and solving that in a separate one, I'm fine with that.
But then instead of making warning, please add an explicit TODO to make it private function after migrating all its users to UntilWithRetry

Reply:
We can't use it in integration tests if that's private...

Reply:
And we shouldn't use that in integration tests.
But as I wrote, I'm fine with TODO to get rid of usages of that later..

Reply:
> And we shouldn't use that in integration tests.

@wojtek-t could you elaborate on that? This PR actually add one valid use case to integration tests `test/integration/apimachinery/watch_restart_test.go`

Reply:
Added a TODO below